### DISCUSSION TITLE
Auth failure using built-in database: "user not found" while the user exists

### DISCUSSION THREAD

**Discussion by marradecki** [emqx/emqx#12709] [Q&A] [answered]
Hello,

Got clueless about the issue below and looking for any hints:

I'm using one-node emqx cluster (image: emqx/emqx:5.5.1) running on k8s. My goal is to use username/password authentication. For this purpose I configured built-in database via dashboard and manually added a user named `telegraf`. Then I configured a TCP listener (1883) and set up authentication with built-in DB. I'm using MQTT client (mqttx) to connect using the username and password. 

It looks like the authentication process starts when MQTT client is connecting, still it fails with "user not found". I believe the user do exists in the database (at least I can see it in the dashboard).

The broker log is:
```
2024-03-14T16:39:38.522998+00:00 [debug] msg: raw_bin_received, mfa: emqx_connection:when_bytes_in/3(781), peername: 10.96.6.1:51748, bin: 102E00044D51545404C2003C000E73696D706C65436C69656E744944000874656C6567726166000874656C6567726166, size: 48, type: hex

2024-03-14T16:39:38.523483+00:00 [debug] msg: mqtt_packet_received, mfa: emqx_channel:handle_in/2(332), peername: 10.96.6.1:51748, clientid: simpleClientID, packet: CONNECT(Q0, R0, D0, ClientId=simpleClientID, ProtoName=MQTT, ProtoVsn=4, CleanStart=true, KeepAlive=60, Username=telegraf, Password=******), tag: MQTT

2024-03-14T16:39:38.523740+00:00 [debug] msg: user_not_found, mfa: emqx_authn_mnesia:authenticate/2(148), peername: 10.96.6.1:51748, clientid: simpleClientID, provider: emqx_authn_mnesia, tag: AUTHN

2024-03-14T16:39:38.523938+00:00 [debug] msg: authenticator_result, mfa: emqx_authn_chains:authenticate_with_provider/2(690), peername: 10.96.6.1:51748, clientid: simpleClientID, authenticator: <<"password_based:built_in_database">>, result: ignore, tag: AUTHN

2024-03-14T16:39:38.524116+00:00 [debug] msg: authentication_result, mfa: emqx_authn_chains:authenticate/2(166), peername: 10.96.6.1:51748, clientid: simpleClientID, reason: chain_result, result: {ok,{error,not_authorized}}, tag: AUTHN

2024-03-14T16:39:38.524347+00:00 [debug] msg: mqtt_packet_sent, mfa: emqx_connection:serialize_and_inc_stats_fun/1(885), peername: 10.96.6.1:51748, clientid: simpleClientID, packet: CONNACK(Q0, R0, D0, AckFlags=0, ReasonCode=5), tag: MQTT

2024-03-14T16:39:38.524793+00:00 [debug] msg: emqx_connection_terminated, mfa: emqx_connection:terminate/2(671), peername: 10.96.6.1:51748, clientid: simpleClientID, reason: {shutdown,not_authorized}, tag: SOCKET

2024-03-14T16:39:38.525031+00:00 [info] msg: terminate, mfa: emqx_connection:terminate/2(676), peername: 10.96.6.1:51748, clientid: simpleClientID, reason: {shutdown,not_authorized}
```

Built-in database users list:
<img width="820" alt="image" src="https://github.com/emqx/emqx/assets/48057331/8fbd8f24-1aab-49de-9f51-e0b1b6954983">
Configuration of the DB:
<img width="822" alt="image" src="https://github.com/emqx/emqx/assets/48057331/d0c4a540-b154-4b0b-9232-489c73d6dcc0">

TCP listener configuration (this is an excerpt from `emqx ctl conf show`):
```
tcp {
    default {
      acceptors = 16
      access_rules = [allow all]
      authentication = [
        {
         backend = built_in_database
         enable = "true"
         mechanism = password_based
         password_hash_algorithm {name = plain, salt_position = disable}
         user_id_type = username
        }
      ]
      bind = 1883
      enable = true
      enable_authn = quick_deny_anonymous
      max_connections = infinity
      mountpoint = ""
      proxy_protocol = false
      proxy_protocol_timeout = 3s
      tcp_options {
        active_n = 100
        backlog = 1024
        buffer = 4KB
        high_watermark = 1MB
        keepalive = none
        nodelay = true
        reuseaddr = true
        send_timeout = 15s
        send_timeout_close = true
      }
    }
```
There's also a global configuration section called `authentication` exactly the same as the one inside listener though I', not sure if this is important.

Abot MQTTX client: connection details are below:
![image](https://github.com/emqx/emqx/assets/48057331/6206dbb7-018c-49ca-b57a-874e70c0fad0)

The db persistence is ON so whenever I restart a pod the user is still there.

Any hints greatly appreciated.

**Comment by ieQu1** [comment]
Hello,

From the logs and the screenshot the configuration looks correct. To help with troubleshooting, can you please run the following command in the emqx shell (it can be entered by running `emqx remote_console` on the host):

```
ets:tab2list(emqx_authn_mnesia).
```
Warning: it will expose the actual password, since you've set password hashing method to `plain`, so make sure to edit it out.

**Reply by marradecki** [reply]
Thank you for the warning, still I'm not afraid about exposing this password since it is a PoC installation and we'll change it later on. The result from the suggestion above is:
```
emqx@processing-emqx-0:/opt/emqx$ emqx remote_console
Erlang/OTP 25 [erts-13.2.2] [emqx] [64-bit] [smp:16:1] [ds:16:1:10] [async-threads:1] [jit:ns]

Restricted Eshell V13.2.2  (abort with ^G)
v5.5.1(processing@processing-emqx-0.processing-emqx-headless.emqx-test.svc.cluster.local)1> ets:tab2list(emqx_authn_mnesia).
[{user_info,{'mqtt:global',<<"telegraf">>},
            <<"telegraf">>,<<>>,true}]
```
Suprising is that in the log it complains about user existence and not bad password. Could it be that we messed up networking configuration (ports) and I'm connecting to a different listener or something? We have a second listener running also:
```
tcp:default
  listen_on       : :1883
  acceptors       : 16
  proxy_protocol  : false
  running         : true
  current_conn    : 0
  max_conns       : infinity
  shutdown_count  : [{not_authorized,1}]
tcp:haproxy
  listen_on       : 0.0.0.0:1884
  acceptors       : 16
  proxy_protocol  : true
  running         : true
  current_conn    : 0
  max_conns       : infinity
  shutdown_count  : []
```

**Comment by zmstone** [comment]
Hi @marradecki 
I noticed that you have authentication enabled per listener.
There is currently no user interface to manage builtin database for per-listener authentication chains.

The authentication chain named `'mqtt:global'` it's the default chain shared auth chain for all listeners.
and this name is stored in the database as "user group".
Since you have `tcp:default` listener enabled with a chain, the chain is assigned with name `tcp:default` (hence used as user group when querying the database).

**Comment by zmstone** [comment]
The code is actually there, but there is no plan to publish it publicly.
Just to demo how it should work (not to encourage the usage of non-public APIs):

I tried to create a json file `a.json`, and import it to the database like below:
```
root@testhost:/opt/emqx# cat a.json
[{"user_id": "a1",
  "password": "bbb"}]
root@testhost:/opt/emqx# emqx eval "emqx_authn_chains:import_users('tcp:default', <<\"password_based:built_in_database\">>, {plain, \"a.json\", element(2,file:read_file(\"a.json\"))})."
ok
```
I can see the records as below:
```
v5.5.1(emqx@192.168.31.216)8> ets:tab2list(emqx_authn_mnesia).
[{user_info,{'tcp:default',<<"a1">>},<<"bbb">>,<<>>,false}]
```
and MQTTX client is able to connect with username `a1` password `bbb`

**Reply by marradecki** [reply]
Thanks for hints. I created `a.json` file and tried to add the user but got an error. No user has been added.
```
$ emqx eval "emqx_authn_chains:import_users('tcp:default', <<\"password_based:built_in_database\">>, {plain, \"a.json\", element(2,file:read_file(\"a.json\"))})."
{error,function_clause}
```
Is there some other way I could add users to `tcp:default` chain?
What I need is username/password auth on one listener (1883). I'm afraid I can't enable authentication for all listeners. My use case is that there's a haproxy listener (port 1884) which gets connections from a haproxy service. The haproxy service terminates SSL and "unpacks" x509 certificates. If a certificate is valid we should accept the connection. Additional auth on haproxy listener would prevent this probably.

**Reply by zmstone** [reply]
I tested my code on the 5.5.1 docker image.
But I would not recommend you to do the same, so I don’t think it worth time looking into the function_clause error.

to meet your requirements, you can configure the haproxy listener to enable_auth=false then it will simply skip the global auth config.

**Comment by marradecki** [comment, marked as answer]
Allright. It seems I got it working by removing specific authentication configuration for the `tcp:default` listener. I just left 
`enable_authn = quick_deny_anonymous` and now I can connect to the listener with username/password as specified in the built-in database. Thank you very much for usefuls suggestions and quick reply!
